src/main/scala/mesosphere/marathon/core/launchqueue/impl/ReviveOffersState.scala

```diff
@@ -4,84 +4,181 @@ package core.launchqueue.impl
+    if (wantsOffers(instance) && hasRecordOfInstanceWantingOffers(instance.role, instance.instanceId)) {
+      this
     } else {
-      copy(scheduledInstances - instance.instanceId, terminalReservations - instance.instanceId)
+      updateInstanceState(instance.role, instance.instanceId, Some(instance))
     }
   }
 
-  /** @return this state with removed ref from [[activeDelays]]. */
+  private def wantsOffers(instance: Instance): Boolean = {
+    instance.isScheduled || shouldUnreserve(instance)
+  }
+
+  private def instanceToWantedInfo(instance: Instance): OffersWantedInfo = {
+    OffersWantedInfo(
+      version,
+      if (shouldUnreserve(instance)) OffersWantedReason.CleaningUpReservations else OffersWantedReason.Launching,
+      instance.runSpec.configRef)
+  }
+
+  /** @return this state with passed instance removed from [[instancesWantingOffers]]. */
+  def withInstanceDeleted(instance: Instance): ReviveOffersState = {
+    updateInstanceState(instance.role, instance.instanceId, Some(instance))
+  }
+
+  /**
+    * Removes delay from the state. If any instance's offers-wanted signal were filtered, then we bump their
+    * version so that a new revive will occur.
+    *
+    * @return this state with removed ref from [[activeDelays]].
+    */
   def withoutDelay(ref: RunSpecConfigRef): ReviveOffersState = {
     logger.debug(s"Marking $ref as no longer actively delayed")
-    copy(activeDelays = activeDelays - ref)
+
+    // This is not optimized
+    val bumpedVersions = instancesWantingOffers.map {
+      case (role, instancesWantedInfo) =>
+        role -> instancesWantedInfo.map {
+          case (instanceId, wantedInfo) =>
+            if (wantedInfo.ref == ref)
+              instanceId -> wantedInfo.copy(version = this.version)
+            else
+              instanceId -> wantedInfo
+        }
+    }
+    copyBumpingVersion(instancesWantingOffers = bumpedVersions, activeDelays = activeDelays - ref)
   }
 
   /** @return this state with updated [[activeDelays]]. */
   def withDelay(ref: RunSpecConfigRef): ReviveOffersState = {
     logger.debug(s"Marking $ref as actively delayed")
-    copy(activeDelays = activeDelays + ref)
+    copyBumpingVersion(activeDelays = activeDelays + ref)
   }
 
-  /** scheduled instances that should be launched. */
-  lazy val scheduledInstancesWithoutBackoff: Set[Instance.Id] =
-    scheduledInstances.view.filter(kv => launchAllowed(kv._2)).map(_._1).toSet
+  /**
+    * Returns a map of all known roles and a desired versioned role-state, where the role-state is OffersWanted or OffersNotWanted.
+    *
+    * A version is used to help indicate whether or not a new revive is wanted for some role; for example
+    *
+    * 1. Offers are indicated as wanted for a role with version 1
+    * 2. The downstream revive directive interprets this as a role should be unsuppressed
+    * 3. Offers are indicated as wanted for a role with version 2
+    * 4. The downstream revive directive issues a new revive
```

Comment:
I don't really follow how is the downstream revive directive interpreting the versions?